lib/aio/s3.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (C) 2022-2024 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+import contextlib
+import hashlib
+import hmac
+import logging
+import mimetypes
+import time
+from types import TracebackType
+from typing import Collection, Mapping, NamedTuple, Self
+
+import aiohttp
+from yarl import URL
+
+from .abc import Destination, LogDriver
+from .jsonutil import JsonError, JsonObject, get_nested, get_str
+from .util import AsyncQueue, create_http_session
+
+logger = logging.getLogger(__name__)
+
+
+class S3Key(NamedTuple):
+    access: str
+    secret: str
+
+
+class HttpRequest(NamedTuple):
+    method: str
+    url: URL
+    headers: Mapping[str, str]
+    data: bytes = b''
+
+
+class HttpQueue:
+    def __init__(self, session: aiohttp.ClientSession, s3_key: S3Key) -> None:
+        self.session = session
+        self._queue = AsyncQueue[HttpRequest]()
+        self._task: asyncio.Task[None] | None = None
+        self._level: int = logging.DEBUG
+        self._s3_key = s3_key
+
+    async def request_once(self, request: HttpRequest, checksum: str) -> None:
+        # NB: Re-sign each attempt.  Time's arrow neither stands still nor reverses.
+        headers = s3_sign(request.url, request.method, request.headers, checksum, self._s3_key)
+        logger.log(self._level, '%s %s', request.method, request.url)
+        async with self.session.request(request.method, request.url, data=request.data, headers=headers) as response:
+            logger.debug('response %s %s %r', request.method, request.url, response)
+
+    async def request(self, request: HttpRequest) -> None:
+        checksum = hashlib.sha256(request.data).hexdigest()
+        # This will keep us retrying for over two minutes before failing
+        for attempt in range(11):
+            try:
+                return await self.request_once(request, checksum)
+            except aiohttp.ClientResponseError as exc:
+                if exc.status < 500:
+                    raise  # We actually care about 4xx errors, but blindly retry 5xx.
+            except aiohttp.ClientError:
+                pass  # That's DNS, connection, etc. errors.  Always retry those.
+
+            # 0.125 → 0.25 → 0.5 → 1 → 4 → 16 → 32s back-off
```

Comment:
This feels too aggressive -- in many cases we got "50x slow down", and retrying after milliseconds is kind of the opposite. Maybe we need to special-case "slow down", but even for DNS or routing errors (overload? congestion?) they have a better chance with less aggressive retrying.

Reply:
I feel like in most cases the error will be resolved by trying again immediately, but if it's not, then we'll slowly back off.

Note that the S3 retry stuff is much more aggressive than GitHub.  We have a much higher limit on S3 (we're paying them, after all)...

There's also no real harm in backing this off either.  Can you make a suggestion about the progression (or algorithm) you'd like to see?

Reply:
Gut feeling: Lower retry to 5, and `sleep(2 ** (2 * attempt)`. That gives 1 → 4 → 16 → 64 → 256. It's not a hard science of course.

Note that this is less about paying, and more about the practictalities of networking :grin:

Reply:
OK.